File: web/public-web/src/repository/__tests__/reportRepository.test.ts
import { beforeEach, describe, expect, it } from 'vitest'
import { createReportRepository } from '../reportRepository'
import { createFakePublicApi } from '../../test/fakePublicApi'
import { _resetForTests, getReport } from '../../storage/db'
import type { ReportDraft } from '../../domain/normalize'
import type { ReportDisplaySnapshot } from '../reportRepository'
import type { PublicApi } from '../../api/publicApi'

const fixedOccurredAt = new Date('2026-01-01T12:00:00Z')
const display: ReportDisplaySnapshot = {
  settlementName: 'Alfaváros',
  railwayLineDisplay: null,
  categoryDisplay: 'Kategória',
  eventTypeDisplay: 'Verekedés',
}

function draft(overrides: Partial<ReportDraft> = {}): ReportDraft {
  return {
    occurredAt: fixedOccurredAt,
    trainIdentifierInput: '  G123  ',
    settlementId: 'settlement-1',
    railwayLineId: null,
    eventTypeCode: 'FIGHT',
    ...overrides,
  }
}

async function resetDb(): Promise<void> {
  await _resetForTests()
  await new Promise<void>((resolve, reject) => {
    const request = indexedDB.deleteDatabase('orszem-public-reports')
    request.onsuccess = () => resolve()
    request.onerror = () => reject(request.error)
    request.onblocked = () => resolve()
  })
}

beforeEach(resetDb)

describe('submit - persist-before-network ordering (§5)', () => {
  it('the local record is committed before the first network attempt', async () => {
    const callLog: string[] = []
    const api: PublicApi = {
      reportCatalog: async () => ({ status: 500, body: undefined, errorBody: undefined }),
      searchSettlements: async () => ({ status: 200, body: [], errorBody: undefined }),
      railwayLinesOfSettlement: async () => ({ status: 500, body: undefined, errorBody: undefined }),
      submitReport: async (body, _accessCredential) => {
        // At the moment the network call fires, the row must already be committed.
        const existing = await getReport(body.clientSubmissionId)
        callLog.push(existing ? 'dao.insert-confirmed-then-api.submitReport' : 'api.submitReport-WITHOUT-prior-insert')
        return {
          status: 201,
          body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
          errorBody: undefined,
        }
      },
      getReport: async () => ({ status: 404, body: undefined, errorBody: undefined }),
    }
    const repository = createReportRepository(api)

    await repository.submit(draft(), display)

    expect(callLog).toEqual(['dao.insert-confirmed-then-api.submitReport'])
  })
})

describe('submit - normalization', () => {
  it('trims the train identifier before sending it', async () => {
    const api = createFakePublicApi()
    api.submitResult = { status: 201, body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' }, errorBody: undefined }
    const repository = createReportRepository(api)

    await repository.submit(draft({ trainIdentifierInput: '  G123  ' }), display)

    expect(api.lastSubmitBody?.trainIdentifier).toBe('G123')
  })

  it('normalizes a blank train identifier to null', async () => {
    const api = createFakePublicApi()
    api.submitResult = { status: 201, body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' }, errorBody: undefined }
    const repository = createReportRepository(api)

    await repository.submit(draft({ trainIdentifierInput: '   ' }), display)

    expect(api.lastSubmitBody?.trainIdentifier).toBeNull()
  })
})

describe('submit - outcomes', () => {
  it('a 201 response marks the record SUBMITTED (Created)', async () => {
    const api = createFakePublicApi()
    const reportId = crypto.randomUUID()
    api.submitResult = { status: 201, body: { reportId, submittedAt: '2026-01-01T12:05:00Z', initialStatus: 'RECEIVED' }, errorBody: undefined }
    const repository = createReportRepository(api)

    const outcome = await repository.submit(draft(), display)

    expect(outcome.kind).toBe('created')
    const stored = await getReport(api.lastSubmitBody!.clientSubmissionId)
    expect(stored?.submissionState).toBe('SUBMITTED')
    expect(stored?.publicReportId).toBe(reportId)
    expect(stored?.publicStatus).toBe('RECEIVED')
  })

  it('an idempotent 200 replay is reported as Replayed, also SUBMITTED', async () => {
    const api = createFakePublicApi()
    api.submitResult = {
      status: 200,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    const repository = createReportRepository(api)

    const outcome = await repository.submit(draft(), display)

    expect(outcome.kind).toBe('replayed')
  })

  it('a thrown network error leaves the record PENDING', async () => {
    const api = createFakePublicApi()
    api.throwOnSubmit = new Error('boom')
    const repository = createReportRepository(api)

    const outcome = await repository.submit(draft(), display)

    expect(outcome.kind).toBe('ambiguous-failure')
    const stored = await getReport(api.lastSubmitBody!.clientSubmissionId)
    expect(stored?.submissionState).toBe('PENDING')
  })

  it('a definitive 400 deletes the record - it is never treated as history', async () => {
    const api = createFakePublicApi()
    api.submitResult = { status: 400, body: undefined, errorBody: { code: 'INVALID_SETTLEMENT', message: '', correlationId: 'x' } }
    const repository = createReportRepository(api)

    const outcome = await repository.submit(draft(), display)

    expect(outcome).toEqual({ kind: 'validation-failed', code: 'INVALID_SETTLEMENT' })
    const stored = await getReport(api.lastSubmitBody!.clientSubmissionId)
    expect(stored).toBeUndefined()
  })

  it('after a 400, a later submission attempt gets a brand new identity', async () => {
    const api = createFakePublicApi()
    api.submitResult = { status: 400, body: undefined, errorBody: { code: 'INVALID_SETTLEMENT', message: '', correlationId: 'x' } }
    const repository = createReportRepository(api)

    await repository.submit(draft(), display)
    const firstId = api.lastSubmitBody!.clientSubmissionId

    api.submitResult = {
      status: 201,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    await repository.submit(draft(), display)
    const secondId = api.lastSubmitBody!.clientSubmissionId

    expect(secondId).not.toBe(firstId)
  })

  it('a 409 marks the record CONFLICT', async () => {
    const api = createFakePublicApi()
    api.submitResult = { status: 409, body: undefined, errorBody: { code: 'IDEMPOTENCY_KEY_REUSED', message: '', correlationId: 'x' } }
    const repository = createReportRepository(api)

    const outcome = await repository.submit(draft(), display)

    expect(outcome.kind).toBe('conflict')
    const stored = await getReport(api.lastSubmitBody!.clientSubmissionId)
    expect(stored?.submissionState).toBe('CONFLICT')
  })
})

describe('retry', () => {
  it('resends the exact same clientSubmissionId, credential and normalized payload', async () => {
    const api = createFakePublicApi()
    api.throwOnSubmit = new Error('boom')
    const repository = createReportRepository(api)

    await repository.submit(draft(), display)
    const firstBody = api.lastSubmitBody
    const firstCredential = api.lastSubmitCredential
    const id = firstBody!.clientSubmissionId

    api.throwOnSubmit = null
    api.submitResult = {
      status: 201,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    await repository.retry(id)

    expect(api.lastSubmitBody).toEqual(firstBody)
    expect(api.lastSubmitCredential).toBe(firstCredential)
  })
})

describe('status refresh', () => {
  it('a generic 404 leaves the record untouched, per the existence-safe contract (§18)', async () => {
    const api = createFakePublicApi()
    api.submitResult = {
      status: 201,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    const repository = createReportRepository(api)
    const submitOutcome = await repository.submit(draft(), display)
    const id = api.lastSubmitBody!.clientSubmissionId

    api.getReportResult = { status: 404, body: undefined, errorBody: { code: 'REPORT_NOT_FOUND', message: '', correlationId: 'x' } }
    const outcome = await repository.refreshStatus(id)

    expect(outcome.kind).toBe('unavailable')
    const stored = await getReport(id)
    expect(stored?.submissionState).toBe('SUBMITTED')
    expect(stored?.publicReportId).toBeDefined()
    expect(submitOutcome.kind).toBe('created')
  })
})

describe('history ordering', () => {
  it('is newest-first', async () => {
    const api = createFakePublicApi()
    api.submitResult = {
      status: 201,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    const repository = createReportRepository(api)
    await repository.submit(draft({ settlementId: 'settlement-a' }), display)
    await new Promise((resolve) => setTimeout(resolve, 5))
    await repository.submit(draft({ settlementId: 'settlement-b' }), display)

    const history = await repository.listHistory()
    const timestamps = history.map((h) => new Date(h.localCreatedAt).getTime())
    expect(timestamps).toEqual([...timestamps].sort((a, b) => b - a))
  })
})

describe('a PENDING record is reloaded correctly by a fresh repository instance (simulating a page reload, §65)', () => {
  it('preserves the frozen payload and retries with the identical identity', async () => {
    const firstApi = createFakePublicApi()
    firstApi.throwOnSubmit = new Error('boom')
    const firstRepository = createReportRepository(firstApi)
    await firstRepository.submit(draft({ trainIdentifierInput: 'G999' }), display)
    const id = firstApi.lastSubmitBody!.clientSubmissionId

    // A fresh repository instance, over the SAME underlying IndexedDB - exactly what a page
    // reload constructs, since the database connection itself is not test-reset here.
    const secondApi = createFakePublicApi()
    secondApi.submitResult = {
      status: 201,
      body: { reportId: crypto.randomUUID(), submittedAt: new Date().toISOString(), initialStatus: 'RECEIVED' },
      errorBody: undefined,
    }
    const secondRepository = createReportRepository(secondApi)

    const reloaded = await secondRepository.listHistory()
    expect(reloaded).toHaveLength(1)
    expect(reloaded[0]?.submissionState).toBe('PENDING')
    expect(reloaded[0]?.trainIdentifier).toBe('G999')

    await secondRepository.retry(id)
    expect(secondApi.lastSubmitBody?.trainIdentifier).toBe('G999')
    expect(secondApi.lastSubmitBody?.clientSubmissionId).toBe(id)
  })
})

File: android/public-app/src/test/kotlin/hu/orszembejelento/app/report/ReportRepositoryTest.kt
package hu.orszembejelento.app.report

import hu.orszembejelento.app.report.data.ReportDisplaySnapshot
import hu.orszembejelento.app.report.data.ReportRepository
import hu.orszembejelento.app.report.data.StatusRefreshOutcome
import hu.orszembejelento.app.report.data.SubmitOutcome
import hu.orszembejelento.app.report.data.network.SubmitReportResponseBody
import hu.orszembejelento.app.report.domain.PublicReportStatus
import hu.orszembejelento.app.report.domain.ReportDraft
import hu.orszembejelento.app.report.domain.SubmissionState
import java.io.IOException
import java.time.Instant
import java.util.UUID
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import retrofit2.Response

class ReportRepositoryTest {

    private val fixedNow = Instant.parse("2026-01-01T12:00:00Z")

    private fun draft(settlementId: UUID = UUID.randomUUID(), trainIdentifier: String? = "  G123  ") = ReportDraft(
        occurredAt = fixedNow,
        trainIdentifierInput = trainIdentifier,
        settlementId = settlementId,
        railwayLineId = null,
        eventTypeCode = "FIGHT",
    )

    private val display = ReportDisplaySnapshot("Alfaváros", null, "Kategória", "Verekedés")

    // ------------------------------------------------------- persist-before-network ordering

    @Test
    fun `the local record is committed before the first network attempt`() = runTest {
        val callLog = mutableListOf<String>()
        val dao = FakeReportHistoryDao(callLog)
        val api = FakePublicApi(callLog).apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(), display)

        assertEquals(listOf("dao.insert", "api.submitReport"), callLog)
    }

    @Test
    fun `a local persistence failure never calls the network at all`() = runTest {
        val api = FakePublicApi()
        val repository = ReportRepository(FakeReportHistoryDao(), FakeCryptoBox(failEncrypt = true), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertEquals(SubmitOutcome.LocalPersistenceFailed, outcome)
        assertEquals(0, api.submitCallCount)
    }

    // -------------------------------------------------------------------------- normalization

    @Test
    fun `the train identifier is trimmed before it is persisted and sent`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(trainIdentifier = "  G123  "), display)

        assertEquals("G123", api.lastSubmitBody?.trainIdentifier)
    }

    @Test
    fun `a blank train identifier normalizes to null`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(trainIdentifier = "   "), display)

        assertNull(api.lastSubmitBody?.trainIdentifier)
    }

    // ------------------------------------------------------------------------------- outcomes

    @Test
    fun `a 201 response marks the record SUBMITTED`() = runTest {
        val dao = FakeReportHistoryDao()
        val publicId = UUID.randomUUID()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(publicId.toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertTrue(outcome is SubmitOutcome.Created)
        val stored = dao.findByClientSubmissionId((outcome as SubmitOutcome.Created).entity.clientSubmissionId)
        assertEquals(SubmissionState.SUBMITTED, stored?.submissionState)
        assertEquals(publicId, stored?.publicReportId)
        assertEquals(PublicReportStatus.RECEIVED, stored?.publicStatus)
    }

    @Test
    fun `an idempotent 200 replay also marks the record SUBMITTED, as Replayed`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(200, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertTrue(outcome is SubmitOutcome.Replayed)
    }

    @Test
    fun `a network failure (thrown exception) leaves the record PENDING`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { throwOnSubmit = IOException("boom") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })
        val settlement = UUID.randomUUID()

        val outcome = repository.submit(draft(settlement), display)

        assertTrue(outcome is SubmitOutcome.AmbiguousFailure)
        val storedRow = requireNotNull(dao.findByClientSubmissionId(findOnlyId(dao)))
        assertEquals(SubmissionState.PENDING, storedRow.submissionState)
    }

    @Test
    fun `a definitive 400 deletes the record - it is never treated as history`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(400, """{"code":"INVALID_SETTLEMENT"}""") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertTrue(outcome is SubmitOutcome.ValidationFailed)
        assertEquals("INVALID_SETTLEMENT", (outcome as SubmitOutcome.ValidationFailed).code)
        assertTrue(dao.observeAll().value.isEmpty())
    }

    @Test
    fun `after a 400, a later submission attempt gets a brand new identity`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(400, """{"code":"INVALID_SETTLEMENT"}""") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(), display)
        val firstIds = dao.observeAll().value.map { it.clientSubmissionId }.toSet()

        api.submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        val second = repository.submit(draft(), display) as SubmitOutcome.Created

        assertFalse(firstIds.contains(second.entity.clientSubmissionId))
    }

    @Test
    fun `a 409 marks the record CONFLICT`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(409, """{"code":"IDEMPOTENCY_KEY_REUSED"}""") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertEquals(SubmitOutcome.Conflict, outcome)
        val storedRow = requireNotNull(dao.findByClientSubmissionId(findOnlyId(dao)))
        assertEquals(SubmissionState.CONFLICT, storedRow.submissionState)
    }

    // ---------------------------------------------------------------------------------- retry

    @Test
    fun `retry resends the exact same clientSubmissionId, credential and normalized payload`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { throwOnSubmit = IOException("boom") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(), display)
        val id = findOnlyId(dao)
        val firstBody = api.lastSubmitBody
        val firstCredential = api.lastSubmitCredential

        api.throwOnSubmit = null
        api.submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        repository.retry(id)

        assertEquals(firstBody, api.lastSubmitBody)
        assertEquals(firstCredential, api.lastSubmitCredential)
    }

    @Test
    fun `retry with an undecryptable credential marks the record ACCESS_LOST and never calls the network`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi()
        // First submit with a working crypto box so a PENDING row exists.
        val workingRepository = ReportRepository(dao, FakeCryptoBox(), api.apply { throwOnSubmit = IOException("boom") }, now = { fixedNow })
        workingRepository.submit(draft(), display)
        val id = findOnlyId(dao)

        val brokenRepository = ReportRepository(dao, FakeCryptoBox(failDecrypt = true), api, now = { fixedNow })
        val callsBeforeRetry = api.submitCallCount
        val outcome = brokenRepository.retry(id)

        assertEquals(SubmitOutcome.AccessLost, outcome)
        assertEquals(callsBeforeRetry, api.submitCallCount)
        assertEquals(SubmissionState.ACCESS_LOST, dao.findByClientSubmissionId(id)?.submissionState)
    }

    // ---------------------------------------------------------------------- status refresh

    @Test
    fun `a generic 404 on status refresh leaves the record untouched, per the existence-safe contract`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })
        val created = repository.submit(draft(), display) as SubmitOutcome.Created

        api.getReportResponse = FakePublicApi.errorResponse(404, """{"code":"REPORT_NOT_FOUND"}""")
        val outcome = repository.refreshStatus(created.entity.clientSubmissionId)

        assertEquals(StatusRefreshOutcome.Unavailable, outcome)
        val stored = dao.findByClientSubmissionId(created.entity.clientSubmissionId)
        assertEquals(SubmissionState.SUBMITTED, stored?.submissionState) // never deleted, never marked access-lost
        assertNotNull(stored?.publicReportId)
    }

    // ------------------------------------------------------- 429: throttled by the backend (B6)

    @Test
    fun `a 429 keeps the record PENDING under the same identity and reports RATE_LIMITED, with no report id`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = FakePublicApi.errorResponse(429, """{"code":"RATE_LIMITED","message":"x","correlationId":"c"}""")
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertEquals(SubmitOutcome.AmbiguousFailure("RATE_LIMITED"), outcome)
        val stored = requireNotNull(dao.findByClientSubmissionId(findOnlyId(dao)))
        assertEquals(SubmissionState.PENDING, stored.submissionState)
        assertEquals("RATE_LIMITED", stored.lastErrorCode)
        assertNull("nothing was created, so there is no report id", stored.publicReportId)
    }

    @Test
    fun `a 429 with no readable body (for example from a proxy) is still recognised as throttling`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(429, "") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })

        val outcome = repository.submit(draft(), display)

        assertEquals(SubmitOutcome.AmbiguousFailure("RATE_LIMITED"), outcome)
        assertEquals(SubmissionState.PENDING, dao.findByClientSubmissionId(findOnlyId(dao))?.submissionState)
    }

    @Test
    fun `a 429 is never retried automatically - exactly one request is made`() = runTest {
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(429, """{"code":"RATE_LIMITED"}""") }
        val repository = ReportRepository(FakeReportHistoryDao(), FakeCryptoBox(), api, now = { fixedNow })

        repository.submit(draft(), display)

        assertEquals("the client must never loop against a limiter", 1, api.submitCallCount)
    }

    @Test
    fun `after a 429 the manual retry resends the exact same identity, credential and payload and can succeed`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(429, """{"code":"RATE_LIMITED"}""") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })
        repository.submit(draft(), display)
        val id = findOnlyId(dao)
        val firstBody = api.lastSubmitBody
        val firstCredential = api.lastSubmitCredential

        val publicId = UUID.randomUUID()
        api.submitResponse = Response.success(201, SubmitReportResponseBody(publicId.toString(), fixedNow.toString(), "RECEIVED"))
        val outcome = repository.retry(id)

        assertTrue(outcome is SubmitOutcome.Created)
        assertEquals("the same clientSubmissionId", firstBody, api.lastSubmitBody)
        assertEquals("the same access credential", firstCredential, api.lastSubmitCredential)
        val stored = requireNotNull(dao.findByClientSubmissionId(id))
        assertEquals(SubmissionState.SUBMITTED, stored.submissionState)
        assertNull("the throttling note is cleared once the report is accepted", stored.lastErrorCode)
        assertEquals("still one record, never a duplicate", 1, dao.observeAll().value.size)
    }

    @Test
    fun `a repeated 429 on retry keeps the record PENDING and keeps saying why`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(429, """{"code":"RATE_LIMITED"}""") }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })
        repository.submit(draft(), display)
        val id = findOnlyId(dao)

        val outcome = repository.retry(id)

        assertEquals(SubmitOutcome.AmbiguousFailure("RATE_LIMITED"), outcome)
        assertEquals(SubmissionState.PENDING, dao.findByClientSubmissionId(id)?.submissionState)
        assertEquals("one call per manual action, no more", 2, api.submitCallCount)
    }

    @Test
    fun `a 429 does not disturb any other outcome - a 503 is still reference-unavailable and other 4xx still retryable`() = runTest {
        val dao503 = FakeReportHistoryDao()
        val api503 = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(503, """{"code":"REFERENCE_DATASET_UNAVAILABLE"}""") }
        assertEquals(
            SubmitOutcome.AmbiguousFailure("REFERENCE_DATASET_UNAVAILABLE"),
            ReportRepository(dao503, FakeCryptoBox(), api503, now = { fixedNow }).submit(draft(), display),
        )
        val api418 = FakePublicApi().apply { submitResponse = FakePublicApi.errorResponse(418, "") }
        val outcome = ReportRepository(FakeReportHistoryDao(), FakeCryptoBox(), api418, now = { fixedNow }).submit(draft(), display)
        assertTrue(outcome is SubmitOutcome.AmbiguousFailure)
    }

    // ------------------------------------------------------------------------------- ordering

    @Test
    fun `observed history is newest-first`() = runTest {
        val dao = FakeReportHistoryDao()
        val api = FakePublicApi().apply {
            submitResponse = Response.success(201, SubmitReportResponseBody(UUID.randomUUID().toString(), fixedNow.toString(), "RECEIVED"))
        }
        val repository = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow.plusSeconds(1) })
        repository.submit(draft(), display)

        val repositoryEarlier = ReportRepository(dao, FakeCryptoBox(), api, now = { fixedNow })
        repositoryEarlier.submit(draft(), display)

        val timestamps = repository.observeHistory().first().map { it.localCreatedAt }
        assertEquals(timestamps.sortedDescending(), timestamps)
    }

    private fun findOnlyId(dao: FakeReportHistoryDao): UUID =
        dao.observeAll().value.single().clientSubmissionId
}
